cmd/branchingconfigmanagers/schedule-generator/main.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/sirupsen/logrus"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/logrusutil"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	crcontrollerutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+
+	"github.com/openshift/ci-tools/pkg/api/ocplifecycle"
+)
+
+type options struct {
+	logLevel          string
+	scheduleDir       string
+	dryRun            bool
+	kubernetesOptions flagutil.KubernetesOptions
+}
+
+func gatherOptions() (*options, error) {
+	o := &options{kubernetesOptions: flagutil.KubernetesOptions{NOInClusterConfigDefault: true}}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	fs.StringVar(&o.logLevel, "log-level", logrus.InfoLevel.String(), "Level at which to log output.")
+	fs.StringVar(&o.scheduleDir, "schedule-dir", "", "Directory holding schedules.")
+	o.kubernetesOptions.AddFlags(fs)
+	fs.BoolVar(&o.dryRun, "dry-run", true, "Do not mutate cluster state.")
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		return o, fmt.Errorf("failed to parse flags: %w", err)
+	}
+	return o, nil
+}
+
+func validateOptions(o *options) error {
+	_, err := logrus.ParseLevel(o.logLevel)
+	if err != nil {
+		return fmt.Errorf("invalid --log-level: %w", err)
+	}
+	return o.kubernetesOptions.Validate(o.dryRun)
+}
+
+func main() {
+	logrusutil.ComponentInit()
+	o, err := gatherOptions()
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to gather options")
+	}
+	if err := validateOptions(o); err != nil {
+		logrus.WithError(err).Fatalf("Invalid options")
+	}
+	level, _ := logrus.ParseLevel(o.logLevel)
+	logrus.SetLevel(level)
+
+	kubeConfigs, err := o.kubernetesOptions.LoadClusterConfigs(nil)
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not load kube config")
+	}
+
+	schedules, err := readSchedules(o.scheduleDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not read schedules.")
+	}
+
+	raw, err := yaml.Marshal(schedules)
+	if err != nil {
+		logrus.WithError(err).Fatal("Could not find marshal schedules")
+	}
+
+	for _, config := range kubeConfigs {
+		client, err := ctrlruntimeclient.New(&config, ctrlruntimeclient.Options{})
+		if err != nil {
+			logrus.WithError(err).Fatal("Could not get client for kube config")
+		}
+		if err := upsertConfigMap(string(raw), client); err != nil {
```

Comment:
So one ugly thing that could happen is CMs going out of sync and then we could act on the wrong config... I think that at least we should continue to the rest of the clusters on errors (we could have already succeeded on some some so aborting does not save much). 

If we want to be robust, we could perhaps mark the CM (via a separate key) as stale before we update it, then update it (and clean the staleness). Config managers would then only need to act on non-stale configs.

Reply:
Interesting idea, but IMHO limited in utility - we would only leave the thing in a stale state if and only if we are able to update it once but not twice.

Reply:
Yeah, we would have only proceed with actual update if we managed to stale CMs everywhere. But I guess operating over stale (but consistent) config everywhere is not good situation either.